### PR TITLE
Image sync

### PR DESCRIPTION
**Pull request by enolfc**
# Summary

Brings image synchronisation to the catchall operations of fedcloud.
Uses atrope for managing images at the sites by generating the configuration for every VO that's avaialable at the AppDB

---

**Related issue :**

### CODE REVIEW

File: deploy/image-sync/CESNET-MCC.tfvars [hunk 1]
```diff
@@ -0,0 +1,8 @@
+# Network
+net_id = "5b8fbb48-461d-4907-ac79-7f7b2ebd17f9"
+
+# Flavor: standard.medium   2cores/4GB RAM
+flavor_id = "4c153ce3-a163-4668-baa7-2cbcb57e2dd8"
+
+# Image: ubuntu bionic
+image_id = "e8d75fc1-ac32-4851-90b5-b4c925e9e6f8"
```
Comment (sebastian-luna-valero) [inline deploy/image-sync/CESNET-MCC.tfvars:8]: Are we using `cloud.egi.eu` VO for this? I don't think `CESNET-MCC` is part of this VO?
Anyway, we shouldn't use `ubuntu bionic`?

File: doc/architecture/decisions/0003-image-synching.md [hunk 2]
```diff
@@ -0,0 +1,47 @@
...
+## Status
+
+Accepted
+
+## Context
+
+EGI provides an image repository (AppDB) for users to share VM images and VOs to
+add them to a list of images to be available at the sites supporting the VO.
+This has been managed with site-level although the uploading of images is for
+most sites a user-level operation that does not need any special privileges.
```
Comment (sebastian-luna-valero) [inline doc/architecture/decisions/0003-image-synching.md:14]: This last sentence is not clear to me.
* `This has been managed with site-level although` seems incomplete
* Uploading of images does require special privileges at the cloud site, right?

File: image-sync/requirements.txt [hunk 3]
```diff
@@ -0,0 +1,4 @@
+git+https://github.com/enolfc/atrope@catchall
```
Comment (sebastian-luna-valero) [inline image-sync/requirements.txt:1]: Note: we should perhaps move this to https://github.com/EGI-Federation/ instead?

**Review by sebastian-luna-valero — COMMENTED** [review]
Many thanks for moving this!

My comments below.

File: deploy/image-sync/CESNET-MCC.tfvars (hunk 1 above)
Comment (enolfc) [reply deploy/image-sync/CESNET-MCC.tfvars:8]: Yes, this is a left over, removed

File: deploy/deploy.sh [hunk 4]
```diff
@@ -4,29 +4,37 @@
...
 
+TMP_SECRETS="$(mktemp)"
 "$PWD/.venv/bin/fedcloud" secret get --locker-token "$FEDCLOUD_SECRET_LOCKER" \
-	deploy data >secrets.yaml
+	deploy data >"$TMP_SECRETS" && mv "$TMP_SECRETS" secrets.yaml
 
-echo "cloud_info_image: \"ghcr.io/egi-federation/fedcloud-cloud-info:sha-$SHORT_SHA\"" >>extra-vars.yaml
+cat >>extra-vars.yaml <<EOF
+cloud_info_image: "ghcr.io/egi-federation/fedcloud-cloud-info:sha-$SHORT_SHA"
+image_sync_image: "ghcr.io/egi-federation/fedcloud-image-sync:sha-$SHORT_SHA"
```
[the remark this answers is not in the document]
Comment (enolfc) [reply deploy/deploy.sh:29]: Yes, this is the commit of the repo, should be ok (although we may need to move image sync and cloud info to different repos if this becomes too complicated)

File: doc/architecture/decisions/0003-image-synching.md (hunk 2 above)
Comment (enolfc) [reply doc/architecture/decisions/0003-image-synching.md]: Updated the text, broken sentence before.
Uploading in default OpenStack does not require privileges, but some EGI sites restrict it. If sites start seeing this as valuable and do not want to widely open the upload of images, we can always create a role for this and ask for mapping it properly 😬 
Please check that it now makes sense.

File: image-sync/requirements.txt (hunk 3 above)
Comment (enolfc) [reply image-sync/requirements.txt:1]: yes, or contribute upstream. There needs to be more changes in atrope to have better caching and management of images if this starts to get traction

File: image-sync/Dockerfile [hunk 5]
```diff
@@ -0,0 +1,54 @@
+FROM python:3-slim AS build
+
+SHELL ["/bin/bash", "-o", "pipefail", "-c"]
+
+# hadolint ignore=DL3008
+RUN apt-get update \
+    && apt-get install -y --no-install-recommends \
+               build-essential gcc git \
```
Comment (sebastian-luna-valero) [inline image-sync/Dockerfile:8]: Do we really need `build-essentials` and `gcc` as dependencies?

File: image-sync/Dockerfile [hunk 6]
```diff
@@ -0,0 +1,54 @@
...
+FROM python:3-slim
+
+LABEL org.opencontainers.image.source=https://github.com/EGI-Federation/fedcloud-catchall-operations
+
+SHELL ["/bin/bash", "-o", "pipefail", "-c"]
+
+# hadolint ignore=DL3015, DL3008
+RUN apt-get update \
+    && apt-get install -y --no-install-recommends \
+       gnupg2 qemu-utils curl \
```
Comment (sebastian-luna-valero) [inline image-sync/Dockerfile:34]: Do we really need `qemu-utils` as dependency?

**Review by sebastian-luna-valero — COMMENTED** [review]
Another couple of comments

File: image-sync/Dockerfile (hunk 6 above)
Comment (enolfc) [reply image-sync/Dockerfile:34]: yes, image conversion needs this

Comment (sebastian-luna-valero) [reply image-sync/Dockerfile:34]: ah, sorry, of course

File: image-sync/Dockerfile (hunk 5 above)
Comment (enolfc) [reply image-sync/Dockerfile]: Changed to regular python image instead of slim so no need to install these

**Review by sebastian-luna-valero — APPROVED** [review]
Great, LGTM!

**Comment by enolfc** [comment]
Merging tomorrow to react to potential failures...

**Pull request merged**